tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,423 @@
+}
+
+Status MemFilesystem::read(
+    const string& path,
+    const uint64_t offset,
+    void* buffer,
+    const uint64_t nbytes) const {
+  FSNode* node = lookup_node(path);
+  if ((node == nullptr) || node->is_dir()) {
+    return LOG_STATUS(Status::MemFSError(
+        string("File not found, read failed for : " + path)));
+  }
+
+  return ((File*)node)->read(offset, buffer, nbytes);
+}
+
+Status MemFilesystem::remove(const string& path, bool is_dir) const {
+  assert(root_);
+  if (path == "") {
+    return LOG_STATUS(
+        Status::MemFSError(string("Cannot remove the root directory")));
+  }
+
+  vector<string> tokens = tokenize(path);
+  auto cur = root_.get();
+  FSNode* parent = nullptr;
+  for (const auto& token : tokens) {
+    if (cur->has_child(token)) {
+      parent = cur;
+      cur = cur->children_[token].get();
+    } else {
+      return LOG_STATUS(Status::MemFSError(
+          string("File not found, remove failed for : " + path)));
+    }
+  }
+
+  if (cur->is_dir() != is_dir) {
+    return LOG_STATUS(
+        Status::MemFSError(string("Remove failed, wrong file type")));
+  }
+
+  if (parent) {
+    parent->children_.erase(tokens.back());
+  }
+  return Status::Ok();
+};
+
+Status MemFilesystem::touch(const string& path) const {
+  assert(root_);
+  vector<string> tokens = tokenize(path);
+  FSNode* cur = root_.get();
+  for (unsigned long i = 0; i < tokens.size() - 1; i++) {
+    auto token = tokens[i];
+    if (!cur->has_child(token)) {
+      cur->children_[token] = unique_ptr<Directory>(new Directory(token));
+    }
+    cur = cur->children_[token].get();
+  }
+
+  string filename = tokens[tokens.size() - 1];
+  cur->children_[filename] = unique_ptr<File>(new File(filename));
+  return Status::Ok();
+}
+
+Status MemFilesystem::write(
+    const string& path, const void* data, const uint64_t nbytes) {
+  FSNode* node = lookup_node(path);
+  /* If the file doesn't exist, create it */
+  if (node == nullptr) {
+    touch(path);
+    node = lookup_node(path);
+    /* If it's a directory, ignore */
+  } else if (node->is_dir()) {
+    return LOG_STATUS(Status::MemFSError(string(
+        "Failed to write on directory, write is only possible on files.")));
+  }
+
+  return ((File*)node)->append(data, nbytes);
+}
+
+vector<string> MemFilesystem::tokenize(const string& path, const char delim) {
+  vector<string> tokens;
+  stringstream ss(path);
+  string token;
+  while (getline(ss, token, delim)) {
+    if (!token.empty()) {
+      tokens.emplace_back(token);
+    }
+  }
+
+  return tokens;
+}
+
+MemFilesystem::FSNode* MemFilesystem::lookup_node(const string& path) const {
+  assert(root_);
+
+  vector<string> tokens = tokenize(path);
+  auto cur = root_.get();
+  for (const auto& token : tokens) {
+    if (cur->has_child(token)) {
+      cur = cur->children_[token].get();
+    } else {
+      return nullptr;
+    }
+  }
+
+  return cur;
+}
+
+MemFilesystem::FSNode* MemFilesystem::lookup_parent_node(
```

Comment:
Let's put a raw-pointer to a parent in the node class and delete this routine. Instead, we'll use `lookup_node` and then get the parent from there.